### PR TITLE
Add yast2_system_settings module for yast2_gui

### PR DESCRIPTION
**Pull request by sofiasyria**
Adding module yast2_system_settings for yast2_gui test scenario

- Related ticket: https://progress.opensuse.org/issues/88361
- Verification runs:  
Tumbleweed
64bit   :  https://openqa.opensuse.org/t1684721
SLE
64bit      :  https://openqa.suse.de/t5742052
ppc64    :  https://openqa.suse.de/t5742053
aarch64 :  https://openqa.suse.de/t5742054

### CODE REVIEW

**Review by jknphy — COMMENTED** [review]
looks promising, I suggested some changes related with renaming which in turns are designing choices that could be improved. 

Optionally, once the review is addressed or any time you want, please consider to use rebase interactive to split your commits, my personal preference of granularity is: 1 for new UI controls (like tab) 1 for controllers and 1 for tests/schedules (but also works 1 for test and 1 for yamls). Besides is more clear to review and revert specific commits if there is a problem a good reason to do it is to not block further work, so anyone can cherry-pick your commits. In particular granularity is important for the ui controls and the controllers.

File: tests/yast2_gui/yast2_system_settings.pm [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+sub validate_sysrq_config {
+    my $expected_status = shift;
+    my %config_value    = (disable => 0,
+        enable => 1);
+    x11_start_program('xterm -geometry 160x45+5+5', target_match => 'xterm');
+    my $is_enabled = script_output("cat /proc/sys/kernel/sysrq");
+    $is_enabled = substr($is_enabled, 0, 1);
+    die "/proc/sys/kernel/sysrq has value $is_enabled , but sysrq should be $expected_status"
+      unless ($is_enabled eq $config_value{$expected_status});
+    $is_enabled = script_output("cat /etc/sysctl.d/70-yast.conf | grep sysrq");
```
[the remark this answers is not in the document]
Comment (sofiasyria) [reply tests/yast2_gui/yast2_system_settings.pm:44]: I haven't considered using this function here, because it works with test data and defined values. In order to use it, I would have to define the values for enabled and disabled sysrq and call the function for different test data, which makes it unnecessarily complicated. The mapping is used because I had issues passing the values 0/1 or true/false as arguments in a function. It's been quite some time so I don't remember all the things I tried. Plus, it makes the comprehension of calling the function very clear. The substr is used because there are occasions where the kernel configuration for enabled sysrq is not just 1 but 126 for example, depending on whether all sysrq functions are allowed or some are not.

File: tests/yast2_gui/yast2_system_settings.pm [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    $self->SUPER::pre_run_hook;
+}
+
+sub validate_sysrq_config {
+    my $expected_status = shift;
+    my %config_value    = (disable => 0,
+        enable => 1);
+    x11_start_program('xterm -geometry 160x45+5+5', target_match => 'xterm');
+    my $is_enabled = script_output("cat /proc/sys/kernel/sysrq");
+    $is_enabled = substr($is_enabled, 0, 1);
```
[the remark this answers is not in the document]
Comment (sofiasyria) [reply tests/yast2_gui/yast2_system_settings.pm:41]: Answered in comment bellow https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/12181#discussion_r603226288

File: lib/YaST/SystemSettings/SystemSettingsController.pm [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    $self->get_system_settings_page->press_ok();
+}
+
+sub remove_pci_id {
+    my ($self) = @_;
+    $self->get_system_settings_page->press_delete();
+    $self->get_system_settings_page->press_ok();
+}
+
+sub config_sysrq {
```
[the remark this answers is not in the document]
Comment (sofiasyria) [reply lib/YaST/SystemSettings/SystemSettingsController.pm:73]: I will change the name to setup_kernel_settings_sysrq but I'm not sure I understand the full proposal. You are proposing that I add a separate function that will just switch tabs? Like: 

sub setup_kernel_settings {
    my ($self) = @_;
    $self->get_system_settings_page->switch_tab_kernel();
}

File: lib/YaST/SystemSettings/SystemSettingsController.pm [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    my ($self, $action) = @_;
+    $self->get_system_settings_page->switch_tab_kernel();
```
[the remark this answers is not in the document]
Comment (sofiasyria) [reply lib/YaST/SystemSettings/SystemSettingsController.pm:75]: I don't like long names :D

File: lib/YaST/SystemSettings/SystemSettingsController.pm [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    $self->{PCIIDSetup}         = YaST::SystemSettings::PCIIDSetup->new({app => YuiRestClient::get_app()});
+    return $self;
+}
+
+sub get_kernel_settings_tab {
+    my ($self) = @_;
+    return $self->{KernelSettingsTab};
+}
+
+sub get_pci_id_setup {
```
[the remark this answers is not in the document]
Comment (sofiasyria) [reply lib/YaST/SystemSettings/SystemSettingsController.pm:43]: This is not a Tab, this is for the pop-up window that the parameters are added.

File: lib/YaST/SystemSettings/SystemSettingsController.pm [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    my $self = bless {}, $class;
+    return $self->init($args);
+}
+
+sub init {
+    my ($self, $args) = @_;
+    $self->{SystemSettingsPage} = YaST::SystemSettings::SystemSettingsPage->new({app => YuiRestClient::get_app()});
+    $self->{ErrorMessage}       = YaST::Warning::ErrorMessage->new({app => YuiRestClient::get_app()});
+    $self->{KernelSettingsTab}  = YaST::SystemSettings::KernelSettingsTab->new({app => YuiRestClient::get_app()});
+    $self->{PCIIDSetup}         = YaST::SystemSettings::PCIIDSetup->new({app => YuiRestClient::get_app()});
```
[the remark this answers is not in the document]
Comment (sofiasyria) [reply lib/YaST/SystemSettings/SystemSettingsController.pm:34]: This is not a Tab, this is for the pop-up window that the parameters are added.

File: lib/YaST/SystemSettings/SystemSettingsPage.pm [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    my ($class, $args) = @_;
+    my $self = bless {
+        app => $args->{app}
+    }, $class;
+    return $self->init();
+}
+
+sub init {
+    my $self = shift;
+    $self->{menu_btn_add} = $self->{app}->menucollection({label => 'Add...'});
```
[the remark this answers is not in the document]
Comment (sofiasyria) [reply lib/YaST/SystemSettings/SystemSettingsPage.pm:29]: I combined it with the PCI ID Setup Tab, but I could separate it.

File: lib/YaST/SystemSettings/SystemSettingsController.pm (hunk 4 above)
[the remark this answers is not in the document]
Comment (jknphy) [reply lib/YaST/SystemSettings/SystemSettingsController.pm:75]: neatpicking but just referring to the order of the words, 'kernel tab' instead of 'tab kernel' it is most common this way.

File: lib/YaST/SystemSettings/SystemSettingsController.pm (hunk 5 above)
[the remark this answers is not in the document]
Comment (jknphy) [reply lib/YaST/SystemSettings/SystemSettingsController.pm:43]: I see, could you please call PCIIDSetupPopup like AcceptLicensePopup, for example

File: lib/YaST/SystemSettings/SystemSettingsController.pm (hunk 3 above)
[the remark this answers is not in the document]
Comment (jknphy) [reply lib/YaST/SystemSettings/SystemSettingsController.pm:73]: just referring to the name of the function, to point to the tab that we are going to setup completely (in this case only one control).

File: lib/YaST/SystemSettings/SystemSettingsPage.pm (hunk 7 above)
[the remark this answers is not in the document]
Comment (jknphy) [reply lib/YaST/SystemSettings/SystemSettingsPage.pm:29]: it would be better, both are doing the same and I understand your approach, the problem potentially could be with those same buttons in both and the transitions. But still is ok, not asking to change it if it works.

File: tests/yast2_gui/yast2_system_settings.pm (hunk 1 above)
[the remark this answers is not in the document]
Comment (jknphy) [reply tests/yast2_gui/yast2_system_settings.pm:44]: anything that is not 0 should mean that is activated, so for Perl should be enough.
This library admit regex, so for checking that an entry starts with '1' should work AFIR.
There is only one test suite, so adding those test data should be fine, you just need to call the function twice and point to the one you need, for example, just a draft of the idea:
```
validate_cfg_file(get_test_suite_data()->{disable_sysrq}->{config_files});
validate_cfg_file(get_test_suite_data()->{enable_sysrq}->{config_files});
```
So basically you call to the function for state A and state B.

File: lib/YuiRestClient/Widget/Tab.pm [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+# Copyright © 2021 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved. This file is offered as-is,
+# without any warranty.
+
+# Maintainer: QE YaST <qa-sle-yast@suse.de>
+
+package YuiRestClient::Widget::Tab;
```
Comment (rwx788) [inline lib/YuiRestClient/Widget/Tab.pm:12]: Could you please put this change into separate commit?

**Review by rwx788 — COMMENTED** [review]
Looks very promising, good job!

File: lib/YuiRestClient/Widget/Tab.pm (hunk 8 above)
Comment (sofiasyria) [reply lib/YuiRestClient/Widget/Tab.pm:12]: Will do. Here I could use menucollection instead of creating new Widget, but I thought the naming is confusing for someone who isn't very familiar, WDYT?

Comment (jknphy) [reply lib/YuiRestClient/Widget/Tab.pm:12]: I prefer tab than dumbtab :) as we did in https://github.com/qe-yast/ruby-yui-rest-client/blob/master/lib/yui_rest_client/widgets/tab.rb

File: schedule/yast/yast2_gui/yast2_gui_sle.yaml [hunk 9]
```diff
@@ -43,6 +43,9 @@ conditional_schedule:
         - yast2_gui/yast2_lan_restart_vlan
         - yast2_gui/yast2_lan_restart_bond
       aarch64:
+        - console/scc_cleanup_reregister
```
[the remark this answers is not in the document]
Comment (sofiasyria) [reply schedule/yast/yast2_gui/yast2_gui_sle.yaml:46]: Because we don't have to do this for s390x, at the moment.

File: tests/yast2_gui/yast2_system_settings.pm [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+sub remove_pci_id {
+    record_info("Remove PCI ID", "Remove the first PCI ID in the table");
+    YaST::Module::open({module => 'system_settings', ui => 'qt'});
+    $system_settings->remove_pci_id();
+}
+
+sub run {
+    my ($self) = shift;
```
[the remark this answers is not in the document]
Comment (sofiasyria) [reply tests/yast2_gui/yast2_system_settings.pm:80]: Ah, thanks. I forgot it there, because I had the content of pre-run hook in run.

File: tests/yast2_gui/yast2_system_settings.pm [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    die "/proc/sys/kernel/sysrq has value $is_enabled , but sysrq should be $expected_status"
+      unless ($is_enabled eq $config_value{$expected_status});
+    $is_enabled = script_output("cat /etc/sysctl.d/70-yast.conf | grep sysrq");
+    $is_enabled = substr($is_enabled, -1, 1);
+    die "sysrq.kernel=$is_enabled in /etc/sysctl.d/70-yast.conf , but sysrq should be $expected_status"
+      unless ($is_enabled eq $config_value{$expected_status});
+    close_xterm();
+}
+
+sub install_package {
```
[the remark this answers is not in the document]
Comment (sofiasyria) [reply tests/yast2_gui/yast2_system_settings.pm:51]: The whole module is designed for gui, so I thought it's not necessary but ok

File: lib/YaST/SystemSettings/SystemSettingsController.pm (hunk 5 above)
[the remark this answers is not in the document]
Comment (sofiasyria) [reply lib/YaST/SystemSettings/SystemSettingsController.pm]: Done

File: lib/YaST/SystemSettings/SystemSettingsController.pm [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+sub remove_pci_id {
+    my ($self) = @_;
+    $self->get_system_settings_page->press_delete();
+    $self->get_system_settings_page->press_ok();
+}
+
+sub config_sysrq {
+    my ($self, $action) = @_;
+    $self->get_system_settings_page->switch_tab_kernel();
+    $self->sysrq_checkbox($action);
```
[the remark this answers is not in the document]
Comment (sofiasyria) [reply lib/YaST/SystemSettings/SystemSettingsController.pm]: Done

File: tests/yast2_gui/yast2_system_settings.pm [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+use y2lan_restart_common 'close_xterm';
+use utils 'zypper_call';
+use x11utils 'start_root_shell_in_xterm';
+
+my $system_settings;
+
+sub pre_run_hook {
+    my ($self) = @_;
+    $system_settings = $testapi::distri->get_system_settings();
+    $self->SUPER::pre_run_hook;
```
Comment (jknphy) [inline tests/yast2_gui/yast2_system_settings.pm:32]: I don't think this line is needed, it is correct though, only thing is the order, always at the end invoke the parent method.

File: tests/yast2_gui/yast2_system_settings.pm (hunk 1 above)
[the remark this answers is not in the document]
Comment (jknphy) [reply tests/yast2_gui/yast2_system_settings.pm:45]: IMO to call the substring from the end and using indexes (second example 70-yast.conf) is not going to be more clear than have fixed test data, with the first example (/proc/sys/kernel/sysrq) might be if you need to get the first character (which is correct but I don't see the need because the range seems to be '0' and '1XX' so Perl booleans should suffice) 

There are several advantages to use libraries, for example you don't need to create code for handle the error.
But if you think that could be out of scope, could you please create a ticket? perhaps there is something to improve in the library to make it more appealing to be used (you are not the first person refusing this suggestion :) )

[the remark this answers is not in the document]
Comment (sofiasyria) [reply tests/yast2_gui/yast2_system_settings.pm:45]: The main reason why I prefer not to change it, is because the current code is ready, tested, working and needs no further effort. I will create a ticket.

File: tests/yast2_gui/yast2_system_settings.pm (hunk 13 above)
Comment (jknphy) [reply tests/yast2_gui/yast2_system_settings.pm:32]: yes, the correct way is to call the parent but as far as I could check we are not doing anything that we need so we can overwrite the method completely removing both lines.

File: lib/Installation/Partitioner/LibstorageNG/v4_3/ErrorDialog.pm [hunk 14]
```diff
@@ -41,4 +41,17 @@ sub press_ok {
     $self->{btn_ok}->click();
 }
 
+sub is_shown {
```
Comment (jknphy) [inline lib/Installation/Partitioner/LibstorageNG/v4_3/ErrorDialog.pm:44]: this belongs to the other commit.

**Review by jknphy — COMMENTED** [review]
LGTM, good job.

File: lib/Installation/Partitioner/LibstorageNG/v4_3/ErrorDialog.pm (hunk 14 above)
Comment (sofiasyria) [reply lib/Installation/Partitioner/LibstorageNG/v4_3/ErrorDialog.pm:44]: done
